tests/test_matrix_exponential.py

```diff
@@ -0,0 +1,136 @@
+import textwrap
+
+import demes
+import numpy as np
+import pytest
+import scipy.linalg
+
+import msprime
+
+
+def _matrix_exponential_eigen(A):
+    """
+    Returns the matrix exponential of A using eigendecomposition.
+    https://en.wikipedia.org/wiki/Matrix_exponential
+    """
+    d, Y = np.linalg.eig(A)
+    Yinv = np.linalg.pinv(Y)
+    D = np.diag(np.exp(d))
+    B = np.matmul(Y, np.matmul(D, Yinv))
+    return np.real_if_close(B, tol=1000)
+
+
+class TestMatrixExponential:
+    """
+    Test cases for the matrix exponential function.
+    """
+
+    def verify(self, A, E=None):
+        assert np.max(np.diag(A)) <= 0
+        assert np.min(A - np.diag(np.diag(A))) >= 0
+        assert np.max(np.sum(A, 1)) <= 0
+        E1 = scipy.linalg.expm(A)
+        E2 = msprime.demography._matrix_exponential(A)
+        E3 = _matrix_exponential_eigen(A)
```

Comment:
since we're comparing to the eigendecomposition, shall we remove the scipy version?